src/Razor/src/Microsoft.AspNetCore.Razor.Language/Intermediate/ComponentAttributeIntermediateNode.cs

```diff
@@ -101,5 +101,42 @@ public override void FormatNode(IntermediateNodeFormatter formatter)
             formatter.WriteProperty(nameof(TagHelper), TagHelper?.DisplayName);
             formatter.WriteProperty(nameof(TypeName), TypeName);
         }
+
+        public bool TryParseEventCallbackTypeArgument(out string argument)
+        {
+            // This is ugly and ad-hoc, but for various layering reasons we can't just use Roslyn APIs
+            // to parse this. We need to parse this just before we write it out to the code generator,
+            // so we can't compute it up front either.
+
+            if (BoundAttribute == null || !BoundAttribute.IsEventCallbackProperty())
+            {
+                throw new InvalidOperationException("This attribute is not an EventCallback attribute.");
+            }
+
+            if (string.Equals(TypeName, ComponentsApi.EventCallback.FullTypeName, StringComparison.Ordinal))
+            {
+                // Non-Generic
+                argument = null;
+                return false;
+            }
+
+            if (TypeName != null && 
+                TypeName.Length > ComponentsApi.EventCallback.FullTypeName.Length + "<>".Length &&
+                TypeName.StartsWith(ComponentsApi.EventCallback.FullTypeName, StringComparison.Ordinal) &&
+                TypeName[ComponentsApi.EventCallback.FullTypeName.Length] == '<' &&
+                TypeName[TypeName.Length - 1] == '>')
```

Comment:
My understanding is that `TypeName` here comes out of Roslyn supplying a canonical representation of the `[Parameter]` type, rather than directly out of any source code written by the developer. Is that right?

I'm assuming therefore that we don't need to worry about things like developers randomly putting spaces between tokens in `Microsoft  . AspNetCore .` etc, and probably don't get affected by type aliasing (`using MyCustomTypeName = EventCallback`).

Please let me know if I'm misunderstanding how this works!

Reply:
That's the original source of the information - it's possible that it could have been tampered with by some other code - but not likely. So basically I'm explaining my hesitation around using such a hack. using a proper parser like Roslyn provides would allow us to detect invalid cases, but this won't. 

I think we should go with this hack for now.